Add multidim-interop test spec (#117)
libp2p/test-plans: multidim-interop/README.md

| muxer        | The muxer to use                                             | no, except when transport is one of quic, quic-v1, webtransport |
| security     | The security channel to use                                  | no, except when transport is one of quic, quic-v1, webtransport |
| is_dialer    | Should you dial or listen                                    | no                                                              |
| test_timeout | Control the timeout of test. In seconds                      | yes, default to 10 seconds                                      |

The test should do two different things depending on if it's the dialer or
listener.

## Dialer

The dialer should emit all diagnostic logs to `stderr`. Only the final JSON
string result should be emitted to `stdout`.

1. Connect to the Redis instance.
2. Create a libp2p node as defined by the environment variables.
3. Get the listener's address via Redis' `BLPOP` using the `listenerAddr` key.
4. Record the current time as `handshakeStart`.
5. Connect to the listener.
6. Ping the listener, and record the round trip time as `pingRTT`
8. Record the time since `handshakeStart`. This is `handshakePlusOneRTT`.

> 9. Print to `stdout` the JSON formatted string: `{"handshakePlusOneRTT":

Reviewer: I assume this is a float in seconds?

   handshakePlusOneRTT, "pingRTT": pingRTT}`.
10. Exit with a code zero.

> On error, the dialer should return a non zero exit code.

Reviewer: ```suggestion
On error, the dialer should return a non-zero exit code.
```

## Listener

The listener should emit all diagnostic logs to `stderr`.

1. Connect to the Redis instance.
2. Create a libp2p node as defined by the environment variables.
3. Publish the listener's address via Redis' `RPUSH` using the `listenerAddr`
   key.
4. Sleep for the duration of test_timeout. The test runner will kill this
   process when the dialer finishes.

> 5. If the timeout is hit, exit with a non-zero error code.

Reviewer: Can’t the server just run indefinitely, until it’s killed by the test runner?

Reviewer: I think this would allow us to get rid of the test timeout env. The timeout would be enforced by the runner, which would simplify the test implementations.

Author: > I think this would allow us to get rid of the test timeout env. 

It wouldn't. You still want a timeout in the redis BLPOP command on the dialer side.

Reviewer: > > I think this would allow us to get rid of the test timeout env.
> 
> It wouldn't. You still want a timeout in the redis BLPOP command on the dialer side.

That timeout doesn’t need to be configurable, does it? There’s no way the redis lookup will take significant time, unless the other host fails to boot.

Author: This is for the `BLPOP` command. This one depends on the other host starting up, listening, and then broadcasting their address. Originally this was hardcoded as 10 seconds, because that seemed like a large enough time frame. But in CI with a NodeJS listener that startup can take more than 10 seconds. This is the increase the spurred this change.

Now we can say, let's just set some arbitrarily high timeout here (e.g. 1 hour) then let the test runner kill things after N seconds. But in my experience, it's been helpful to see where the tests fails with a timeout. Did the dialer fail to get the listener's address? Or did the dialer fail to connect and ping?

# Interoperability test

This tests that different libp2p implementations can communicate with each other
on each of their supported capabilites.

Each version of libp2p is defined in `versions.ts`. There the version defines
its capabilities along with the id of its container image.

# Test spec

The implementation is run in a container and is passed parameters via
environment variables. The current parameters are:

| Name         | Description                                                  | Is Optional                                                     |
| ------------ | ------------------------------------------------------------ | --------------------------------------------------------------- |

> | REDIS_ADDR   | A different address to connect to redis (default redis:6379) | yes                                                             |

Reviewer 2: ```suggestion
| REDIS_ADDR   | A different address to connect to redis (default redis:6379) | yes, when not provided implementations should default to the `redis` host on port 6379
|
```

| ip           | IP address to bind the listener to                           | no                                                              |

> | transport    | The transport to use                                         | no                                                              |

Reviewer 3: Is it useful to define the valid values for "transport"?  (same goes for "muxer" and "security").

Do we need to define the behavior if it gets pass a "transport" that it doesn't have a mapping for?

Author: Valid values of transport depend on what the implementation says it supports. e.g. if an implementation doesn't support TCP, it'll never be called with a `transport=tcp` param.

| muxer        | The muxer to use                                             | no, except when transport is one of quic, quic-v1, webtransport |
| security     | The security channel to use                                  | no, except when transport is one of quic, quic-v1, webtransport |
| is_dialer    | Should you dial or listen                                    | no                                                              |

> | test_timeout | Control the timeout of test. In seconds                      | yes, default to 10 seconds                                      |

Reviewer 3: ```suggestion
| test_timeout_seconds | Control the timeout of test.                     | yes, default to 10 seconds                                      |
```

Make self documenting?

The test should do two different things depending on if it's the dialer or
listener.

## Dialer

The dialer should emit all diagnostic logs to `stderr`. Only the final JSON
string result should be emitted to `stdout`.

1. Connect to the Redis instance.
2. Create a libp2p node as defined by the environment variables.
3. Get the listener's address via Redis' `BLPOP` using the `listenerAddr` key.

> 4. Record the current time as `handshakeStart`.

Reviewer 3: I assume there is "time" terminology we can standardize on?  (I don't know the latest and greatest.  I assume there are normal conventions on this.  My default is to use the precise terms back from jodatime (https://www.joda.org/joda-time/ ) which I believe are in other languages/libraries.)

```suggestion
4. Record the current instant as `handshakeStartInstant`.
```

Just an idea for clarity.  Your call of course.

5. Connect to the listener.

> 6. Ping the listener, and record the round trip time as `pingRTT`

Reviewer 3: ```suggestion
6. Ping the listener, and record the duration of the round trip as `pingRoundTripDuration`
```

(idea for clarity to avoid the ambiguous term "time".)

Author: RTT is a well know term. I'll keep it.

> 8. Record the time since `handshakeStart`. This is `handshakePlusOneRTT`.

Reviewer 3: ```suggestion
8. Record the duration since `handshakeStartInstance`. This is `handshakePlusOneRoundTripDuration`.
```

> 9. Print to `stdout` the JSON formatted string: `{"handshakePlusOneRTT":

Reviewer 3: ```suggestion
9. Print to `stdout` the JSON formatted string: `{"handshakePlusOneRoundTripDurationMillis":
```

Going for clarify on this being a duration and the units.

>    handshakePlusOneRTT, "pingRTT": pingRTT}`. Durations should be printed in

Reviewer 3: ```suggestion
   handshakePlusOneRoundTripDuration, "pingRoundTripDurationMillis": pingRoundTripDuration}`. Durations should be printed in
```

   milliseconds as a float.
10. Exit with a code zero.

On error, the dialer should return a non-zero exit code.

## Listener

The listener should emit all diagnostic logs to `stderr`.

1. Connect to the Redis instance.
2. Create a libp2p node as defined by the environment variables.
3. Publish the listener's address via Redis' `RPUSH` using the `listenerAddr`
   key.

> 4. Sleep for the duration of test_timeout. The test runner will kill this

Reviewer 3: ```suggestion
4. Sleep for the duration of `test_timeout_seconds`. The test runner will kill this
```
